### PR TITLE
Logical right and left shift diagnostics.

### PR DESCRIPTION
**Pull request by moonheart08**
This adds lints for `<<<` and `>>>`. They explicitly do not suggest replacements as I could not figure out a simple way to do so for `>>>`, and for `<<<` the replacement is subjective (it could be `<<`, but that doesn't preserve sign.)

### CODE REVIEW

File: src/test/ui/operator-recovery/arithmetic-right-shift.stderr [hunk 1]
```diff
@@ -0,0 +1,8 @@
+error: invalid shift operator `>>>`
+  --> $DIR/arithmetic-right-shift.rs:2:22
+   |
+LL |     println!("{}", 1 >>> 2);
+   |                      ^^^ `>>>` is not a valid right shift operator, consider casting to a signed integer and right shifting normally.
```
Comment (scottmcm) [inline src/test/ui/operator-recovery/arithmetic-right-shift.stderr:5]: I've never used `>>>` so I might be confused, but isn't `>>>` usually a logical shift (that brings in zeros), thus one should cast to an **un**signed integer?

That's what it seems to be in https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Unsigned_right_shift and https://docs.oracle.com/javase/specs/jls/se7/html/jls-15.html#jls-15.19, at least.

Comment (moonheart08) [reply src/test/ui/operator-recovery/arithmetic-right-shift.stderr:5]: Yoooou are correct, I mixed the two up in my head. I'll fix it in a few minutes.

File: compiler/rustc_error_messages/locales/en-US/parser.ftl [hunk 2]
```diff
@@ -46,6 +46,10 @@ parser_invalid_comparison_operator = invalid comparison operator `{$invalid}`
     .use_instead = `{$invalid}` is not a valid comparison operator, use `{$correct}`
     .spaceship_operator_invalid = `<=>` is not a valid comparison operator, use `std::cmp::Ordering`
 
+parser_invalid_shift_operator = invalid shift operator `{$invalid}`
+    .arithmetic_left_shift_operator_invalid = `<<<` is not a valid left shift operator, consider shifting normally and fixing the sign as needed.
+    .arithmetic_right_shift_operator_invalid = `>>>` is not a valid right shift operator, consider casting to a signed integer and right shifting normally.
```
Comment (estebank) [inline compiler/rustc_error_messages/locales/en-US/parser.ftl:51]: We don't end sentences with period (nor we capitalize the first letter in sentences)

```suggestion
    .arithmetic_left_shift_operator_invalid = `<<<` is not a valid left shift operator, consider shifting normally and fixing the sign as needed
    .arithmetic_right_shift_operator_invalid = `>>>` is not a valid right shift operator, consider casting to an unsigned integer and right shifting normally
```

I would prefer if the `consider...` text was instead a `note`, separate from the main label

File: compiler/rustc_parse/src/parser/expr.rs [hunk 3]
```diff
@@ -279,6 +280,35 @@ impl<'a> Parser<'a> {
...
+                self.sess.emit_err(InvalidShiftOperator {
+                    span: sp,
+                    invalid: ">>>".into(),
+                    sub: InvalidShiftOperatorSub::ArithmeticRightShift(sp),
+                });
+                self.bump();
+            }
+
+            // Look for the arithmetic left shift operator (`<<<`) and recover.
+            // Yea nobody uses this but it's worth being thorough.
```
Comment (davidtwco) [inline compiler/rustc_parse/src/parser/expr.rs:298]: I don't think this second line in the comment is worth keeping.

File: compiler/rustc_error_messages/locales/en-US/parser.ftl (hunk 2 above)
Comment (davidtwco) [reply compiler/rustc_error_messages/locales/en-US/parser.ftl:51]: I agree with both of these comments.

**Comment by Nilstrieb** [comment]
Hi, I've seen you changed some diagnostic structs in your PR. After #103345, the way we refer to fluent messages changed. They are now in a flat namespace with the same identifier as in the fluent file. For example,  `parser::cool_thing` is now `parser_cool_thing` and `parser::suggestion` just `suggestion`.
You should [rebase](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts) to the latest master and change your fluent message references as described above. Thanks!

**Comment by moonheart08** [comment]
Ah shoot, this slipped under my radar. I'll put finishing this on my calender.

**Comment by apiraino** [comment]
Switching to waiting on author to incorporate changes. Feel free to request a review with `@rustbot ready`, thanks!

@rustbot author

**Comment by Dylan-DPC** [comment]
@moonheart08 any updates on this?

**Comment by Dylan-DPC** [comment]
Closing this as inactive. Feel free to reöpen this pr or create a new pr if you get the time to work on this. Thanks

**Comment by Dylan-DPC** [comment]
l

**Pull request closed without merging**